### ISSUE TITLE
Suggest transducers for long seq chains

### ISSUE THREAD

**Issue by thenonameguy** [borkdude/clj-kondo#323]
To improve performance and reduce GC load in our applications it would be preferable to use transducers in longer `seq` generating chains.

```clojure
(->> my-seq
     (map inc)
     (filter odd?)
     (map inc)
     (reduce +))
```
clj-kondo could suggest
```clojure
(transduce 
  + 
  (comp
   (map inc)
   (filter odd?)
   (map inc)) 
  my-seq)

;; or at least
(reduce + (sequence (comp (map inc) (filter odd?) (map inc)) my-seq))

```
This would require knowing at least for `clojure.core` which functions consume/produce seq's and have transducer versions. (use spec fdefs?)

**Comment by borkdude** [comment]
@thenonameguy 

This seems like a cool feature, but there might be certain trade-offs to be made.
I wrote a `cond` -> `case` linter a while ago:
<img width="226" alt="cond-as-case" src="https://user-images.githubusercontent.com/284934/61800600-9d759180-ae2d-11e9-91da-8dd10f4026ba.png">
but then @puredanger made a good point that case is not always faster than `cond`. The same might be true for transducers vs lazy collections?

**Comment by thenonameguy** [comment]
I think your linter as well as my proposal would be a useful feature even in presence of false positive cases. It would be fine to keep as them as an INFO level linter, that way more novice devs get the chance to learn compared to the current state. Transducers are severely underutilized today IMO, just because most devs are used to dealing with sequence abstractions directly and don't even consider them in day-to-day coding. Anyway, since the implementation complexity for this is relatively high, this could be kept as an idea issue until all low-hanging fruit has been reaped for clj-kondo.

Thanks again for this awesome tool! 🎉

**Comment by andreyorst** [comment]
I've wrote this hook to address this problem:

```clojure
(ns hooks.thread-macros
  (:require [clj-kondo.hooks-api :as api]))

(def consecutive-threshold 2)
(def transduceable #{'map
                     'filter})

(defn consecutive-threading [{:keys [:node]}]
  (let [[_ & forms] (rest (:children node))]
    (loop [forms forms
           counter 0]
      (when-let [form (first forms)]
        (if (contains? transduceable (some-> form :children first api/sexpr))
          (if (>= counter consecutive-threshold)
            (api/reg-finding!
             (assoc (meta node)
                    :message "prefer transducers to consecutive maps or filters in threading macros"
                    :type :expensive-threading))
            (recur (next forms)
                   (inc counter)))
          (recur (next forms)
                 0))))))
```

It walks the tree of `->>` macro counting sequential calls to `transduceable` functions, and reports if more than 2 were found. Although not every form of threading can be turned into equivalent transducer composition, but I've never seen this hook triggering on such patterns.

As pointed out - it is not always feasible to warn about this, as user may explicitly want a lazy collection as a result, and it aldo doesn't cover macroexpanded form, e.g. when we literally use `(map f1 (filter p (map f col)))` without threading macro, though I've never seen this, and usually see thread macro used in the code.

**Comment by borkdude** [comment]
Cool that you could write your own additional linting in a hook!

**Comment by andreyorst** [comment]
> Cool that you could write your own additional linting in a hook!

This is actually my favourite part about clj-kondo. Coming from C/C++ where linters are quite smart, yet can't be configured to suit the domain I'm working in, being able to do it for Clojure is a relief. This really helps conveying code style to other members of the team.

**Comment by borkdude** [comment]
I think the hook based approach by @andreyorst is the way to go forward with this.
